machine/netberg/netberg_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch

```diff
@@ -0,0 +1,278 @@
++    char filename[20] = "";
++
++    sprintf(filename, "/dev/i2c-%d", i2cbus);
++    file = open(filename, O_RDWR);
++
++    if (file < 0)
++        printf("Error: Could not open file %s\n", filename);
++
++    return file;
++}
++
++static inline __s32 i2c_dev_read(__s32 file, __u8 *buffer, __s32 offset, __s32 length)
++{
++    __s32 i;
++    union i2c_smbus_data data;
++
++    for (i = 0; i < length; i++)
++    {
++        if (i2c_dev_smbus_access(file, I2C_SMBUS_READ, (__u8)(i + offset), I2C_SMBUS_BYTE_DATA, &data))
++        {
++            return -1;
++        }
++        else
++        {
++            buffer[i] = 0x0FF & data.byte;
++        }
++        usleep(5000);
++    }
++
++    return 0;
++}
++
++static inline __s32 i2c_dev_write(__s32 file, __u8 *buffer, __s32 offset, __s32 length)
++{
++    __s32 i;
++    union i2c_smbus_data data;
++
++    for (i = 0; i < length; i++)
++    {
++        data.byte = (__u8)buffer[i];
++        if (i2c_dev_smbus_access(file, I2C_SMBUS_WRITE, (__u8)(i + offset), I2C_SMBUS_BYTE_DATA, &data) != 0)
++            return -1;
++        usleep(5000);
++    }
++    return 0;
++}
++
++static inline __s32 i2c_dev_seek(__s32 file, __s32 offset)
++{
++    return i2c_dev_smbus_access(file, I2C_SMBUS_WRITE, offset, I2C_SMBUS_BYTE, NULL);
++}
++
++static inline __s32 i2c_read(__s32 bus, __u8 dev, uint addr, __u8 *data, __s32 length)
++{
++    __s32 fd, ret;
++
++    fd = i2c_dev_open(bus);
++    if(fd < 0) {
++        return -1;
++    }
++
++    ret = ioctl(fd, I2C_SLAVE, dev);
++    if(ret < 0) {
++        close(fd);
++        return -1;
++    }
++
++    ret = i2c_dev_read(fd, data, addr, length);
++    if(ret < 0) {
++        close(fd);
++        return -1;
++    }
++
++    close(fd);
++    return 0;
++}
++
++static inline __s32 i2c_write(__s32 bus, __u8 dev, uint addr, __s32 alen, __u8 *data, __s32 length)
++{
++    __s32 fd, ret;
++
++    fd = i2c_dev_open(bus);
++    if(fd < 0) {
++        return -1;
++    }
++
++    ret = ioctl(fd, I2C_SLAVE, dev);
++    if(ret < 0) {
++        close(fd);
++        return -1;
++    }
++
++    ret = i2c_dev_seek(fd, addr);
++    if(ret < 0) {
++        close(fd);
++        return -1;
++    }
++
++    ret = i2c_dev_write(fd, data, alen, length);
++    if(ret < 0) {
++        close(fd);
++        return -1;
++    }
++
++    close(fd);
++    return 0;
++}
++
++/**
++ * i2c_mux_select - select channel for i2c mux
++ * 0xff will release i2c mux
++ */
++static inline __s32 i2c_mux_select(__s32 bus, __u8 chip, __u8 channel)
++{
++    __u8 data = 0;
++    if(channel == 0xff)
++    {
++        data = 0;
++    }
++    else if(channel >=0 && channel <8)
++    {
++        data = (0x01 << channel);
++    }
++    else
++    {
++        return -1;
++    }
++
++    return i2c_write(bus, chip, 0, 1, &data, 1);
++}
++
++/**
++ * i2c_mux_release - release channel for i2c mux
++ */
++static inline __s32 i2c_mux_release(__s32 bus, __u8 chip)
++{
++    return i2c_mux_select(bus, chip, 0xff);
++}
```

Comment:
It cannot be simpler. Because before reading or writing EEPROM, the channel on the I2C-bus switch must be selected.